File: 295. Find Median from Data Stream/src/Main.java
import java.util.PriorityQueue;

public class Main {
    public static void main(String... args){
        MedianFinder obj = new MedianFinder();
        obj.addNum(1);
        obj.addNum(2);
        System.out.println(obj.findMedian());
        obj.addNum(3);
        System.out.println(obj.findMedian());

        /*PriorityQueue<Integer> smaller;
        PriorityQueue<Integer> larger;
        smaller = new PriorityQueue<>();
        larger = new PriorityQueue<>((a, b) ->
                b-a
        );

        smaller.offer(1);
        smaller.offer(2);
        smaller.offer(3);

        larger.offer(1);
        larger.offer(2);
        larger.offer(3);

        while (!smaller.isEmpty()){
            System.out.println(smaller.poll());
        }

        while (!larger.isEmpty()){
            System.out.println(larger.poll());
        }*/
    }
}

File: 295. Find Median from Data Stream/main.py
import heapq

class MedianFinder:

    def __init__(self):
        self.accending = []
        self.decending = []

    def addNum(self, num: int) -> None:
        sizeA, sizeD = len(self.accending), len(self.decending)
        if sizeD >= sizeA:
            temp = heapq.heappushpop(self.decending, -num)
            heapq.heappush(self.accending, -temp)
        else:
            temp = heapq.heappushpop(self.accending, num)
            heapq.heappush(self.decending, -temp)

    def findMedian(self) -> float:
        sizeA, sizeD = len(self.accending), len(self.decending)
        if sizeA < sizeD:
            return -heapq.nsmallest(1, self.decending)[0]
        elif sizeD < sizeA:
            return heapq.nsmallest(1, self.accending)[0]
        else:
            return (-heapq.nsmallest(1, self.decending)[0] + heapq.nsmallest(1, self.accending)[0]) / 2

m = MedianFinder()
m.addNum(-1)
print(m.findMedian())
m.addNum(-2)
print(m.findMedian())
m.addNum(-3)
print(m.findMedian())
m.addNum(-4)
print(m.findMedian())
m.addNum(-5)
print(m.findMedian())




# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
